alphagov/paas-cf-conduit
# [#155431687] Support Redis CLI

## What

I added support for Redis CLI.

Additional changes:
 - support for redis-cli (we add extra args to the redis-cli command to set host, port and password)
 - if the redis service is using TLS and the redis-cli program is used we start a TLS tunnel so a non-TLS enabled client, like redis-cli can connect
 - big refactor, all logic from the command was moved to a separate package and struct
 - several structs were moved to separate packages (out from main)
 - the credentials structure was changed to map[string]interface{} instead of a fix structure
 - all VCAP_SERVICES fields will be automatically updated with the local address

Bonus:
 - we check if a known application has a matching service type passed (so you get an error if you try to connect to a Postgres db with the mysql client), fixes https://github.com/alphagov/paas-cf-conduit/issues/32

❗️ Before merging bump the minor version in https://github.com/alphagov/paas-cf-conduit/blob/fbcbbf65c51c9945ce2b82d41735df3a818824e9/plugin.go#L65

## How to review

1. Create a service instance for all service types

```
cf create-service mysql Free mysql-test
cf create-service postgres Free postgres-test
cf create-service redis tiny-unclustered redis-test
```

1. Connect to all services with all known clients

```
cf conduit mysql-test -- mysql
cf conduit mysql-test -- mysqldump

cf conduit postgres-test -- psql
cf conduit postgres-test -- pgdump

cf conduit redis-test -- redis-cli
cf conduit redis-test -- redis-cli set foo bar
```

1. Check if connecting to the wrong service throws an error

```
cf conduit redis-test -- mysql
```

1. Check if all environment variables are properly set (all connection params should be updated with localhost and local port in VCAP_SERVICES)

```
cf conduit mysql-test -- env
cf conduit postgres-test -- env
cf conduit redis-test -- env
```

## Who can review

Not me.
